File: Interview150/Matrix/54.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  vector<int> spiralOrder(vector<vector<int>> &matrix) {
    if (matrix.empty())
      return {};
    vector<int> spiral;
    int top = 0, bottom = matrix.size() - 1;
    int left = 0, right = matrix[0].size() - 1;

    while (top <= bottom && left <= right) {
      for (int j = left; j <= right; ++j)
        spiral.push_back(matrix[top][j]);
      top++;

      for (int i = top; i <= bottom; ++i)
        spiral.push_back(matrix[i][right]);
      right--;

      if (top <= bottom) {
        for (int j = right; j >= left; --j)
          spiral.push_back(matrix[bottom][j]);
        bottom--;
      }

      if (left <= right) {
        for (int i = bottom; i >= top; --i)
          spiral.push_back(matrix[i][left]);
        left++;
      }
    }
    return spiral;
  }
};

int main() {
  Solution s;
  vector<vector<int>> matrix1 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
  vector<int> ret = s.spiralOrder(matrix1);
  for (auto num : ret) {
    cout << num << " ";
  }
}

File: Interview/Databricks/Leetcode/30d/54.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n, m, x, y, idx = len(matrix), len(matrix[0]), 0, 0, 0
        if n == m and m == 1:
            return [1]
        dir = [(0, 1), (1, 0), (0, -1), (-1, 0)] # E, S, W, N
        ret = []

        def inside(x, y):
            return x < n and x >= 0 and y < m and y >= 0
        
        while True:
            print(x, y)
            ret.append(matrix[x][y])
            matrix[x][y] = -105

            nx, ny = x + dir[idx][0], y + dir[idx][1]
            if not inside(nx, ny) or matrix[nx][ny] == -105:
                idx = (idx + 1) % 4
            else:
                x, y = nx, ny
                continue
            
            nx, ny = x + dir[idx][0], y + dir[idx][1]
            if not inside(nx, ny) or matrix[nx][ny] == -105:
                break
            
            x, y = nx, ny
        
        return ret